remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -1391,45 +1402,96 @@ where
-                .finish_write = true;
+                UploadStatus::Ongoing(OngoingUploadStatus::Done) => {
+                    // Another task has already completed the upload of this artifact, no need to do any work.
+                }
+                UploadStatus::Ongoing(OngoingUploadStatus::Error) => {
+                    // Another task tried to perform the transmission, but failed.
+                    anyhow::bail!("Upload queued for previous action failed.")
+                }
+                UploadStatus::New(tx) => {
+                    let mut file = tokio::fs::File::open(&name)
+                        .await
+                        .with_context(|| format!("Opening `{name}` for reading failed"))?;
+                    let mut data = vec![0; max_total_batch_size];
+
+                    let mut write_offset = 0;
+                    let mut upload_segments = Vec::new();
+                    loop {
+                        let length = file
+                            .read(&mut data)
+                            .await
+                            .with_context(|| format!("Error reading from {name}"))?;
+                        if length == 0 {
+                            break;
+                        }
+                        upload_segments.push(WriteRequest {
+                            resource_name: resource_name.to_owned(),
+                            write_offset,
+                            finish_write: false,
+                            data: data[..length].to_owned(),
+                        });
+                        write_offset += length as i64;
+                    }
+                    upload_segments
+                        .last_mut()
+                        .with_context(|| format!("Read no segments from `{name} "))?
+                        .finish_write = true;
+
+                    let upload_ret = bystream_fut(upload_segments)
+                        .await
+                        .and_then(|resp| {
+                            if resp.committed_size != digest.size_in_bytes {
+                                Err(anyhow::anyhow!(
+                                    "Failed to upload `{name}`: invalid committed_size from WriteResponse"
+                                ))
+                            }
+                            else {
+                                Ok(())
+                            }
+                        });
 
-            let resp = bystream_fut(upload_segments).await?;
-            if resp.committed_size != size {
-                return Err(anyhow::anyhow!(
-                    "Failed to upload `{name}`: invalid committed_size from WriteResponse"
-                ));
+                    // Mark artifact as uploaded and notify other potentially waiting tasks.
+                    if upload_ret.is_ok() {
+                        prev_uploads.alter(&digest, |_, _| OngoingUploadStatus::Done);
```

Comment:
does prev_uploads grown unbounded?  was expecting to see a removal here to keep it sized to the inflight digests

Reply:
My concern was with the following race:
1. Task 1: started an upload of digest ABC
2. Task 2: finds ABC is missing as the upload hasn't succeeded yet
3. Task 1: finishes upload, notifies other ongoing uploads of success, removes entry from `prev_uploads`
4. Task 2: starts upload, causing it to attempt to reupload

Thinking about this more the issue isn't that bad, as RE should immediately respond to the upload request that it is fully committed. To remove that final upload we could also directly populate `REClient::find_missing_cache` once the upload is finished and checking `REClient::find_missing_cache` before starting an upload.